### ISSUE TITLE
Revised Cicero parser

### ISSUE THREAD

**Issue by jeromesimeon** [accordproject/cicero#547]
**Is your feature request related to a problem? Please describe.**
The current Cicero parser is fairly hard to maintain, leads to performance problems (#523), and is difficult to make extensible. It also doesn't know about markdown (notably is not aware of whitespace rules) which can lead to confusing errors for users.

**Describe the solution you'd like**
A full revision, possibly including switching to a separate parsing technology, which would have the following properties:
- Expressive enough for our use cases
- Lightweight (small library and memory footprint during parsing)
- Composable to facilitate extensions
- Can be used in the browser

**Describe alternatives you've considered**
There are lots of parsing technologies available, with different trade-offs. Some of the options:
- Stick to generate a parser with Nearley
- Switch to another parser generator (e.g., pegjs or antlr)
- Switch to a different kind of parsing technology (e.g., regex!? or maybe more realistically a parser combinator library like https://github.com/jneen/parsimmon)

**Additional context**

**Comment by jeromesimeon** [comment]
Backbone for a new parser based on the Parsimmon library is available at: https://github.com/accordproject/markdown-transform/tree/js-template-parse

**Comment by jeromesimeon** [comment]
Checklist:
- [ ] Expressivity
   - [ ] Text blocks
   - [ ] Variables
       - [ ] Double
       - [ ] String
       - [ ] Integer
       - [ ] DateTime
   - [ ] Blocks
       - [ ] Conditional blocks
       - [ ] With blocks
       - [ ] Clause blocks
       - [ ] List blocks
- [ ] Performances
       - [ ] Runtime
       - [ ] Memory footprint
- [ ] Browser support

**Comment by jeromesimeon** [comment]
Added a test with a "large" document (it's not that large, only about 300k, but that's about 1600 lines and essentially based on the "huge" document in the Slate demo).

Parsing time (including error) looks pretty good so far:
```
bash-3.2$ du -k test/data/large.md 
284	test/data/large.md
bash-3.2$ du -k test/data/largeErr.md 
284	test/data/largeErr.md
bash-3.2$ time ./bin/index.js parse --template ./test/data/templateLarge.json --input ./test/data/large.md 
4:25:09 PM - info:
{
  "$class": "org.test.MyContract",
  "clause3": {
    "$class": "org.test.MyClause",
    "seller": "Steve",
    "buyer": "Betty",
    "amount": 3131,
    "currency": "EUR",
    "forceMajeure": false
  },
  "penalty": 10.99
}
./bin/index.js parse --template ./test/data/templateLarge.json --input   0.22s user 0.05s system 104% cpu 0.260 total
bash-3.2$ time ./bin/index.js parse --template ./test/data/templateLarge.json --input ./test/data/largeErr.md 
4:25:51 PM - error: {"status":false,"index":{"offset":145271,"line":803,"column":40},"expected":["/\"[^\"]*\"/"]}
./bin/index.js parse --template ./test/data/templateLarge.json --input   0.21s user 0.05s system 104% cpu 0.249 total
bash-3.2$ 
```

**Comment by jeromesimeon** [comment]
Pretty good news on the memory usage front (and I believe excellent news on the runtime performances).

For a small template: memory usage shot at 3M (most of which is code)
<img width="1519" alt="Screenshot 2020-04-23 at 5 30 29 PM" src="https://user-images.githubusercontent.com/670099/80151688-74cf2380-8588-11ea-86d0-65bdc3ee6241.png">

For a 300k template: memory usage shoots up to 10M (the extra memory seems to be mostly static strings). It's a fairly rough/quick estimate.
<img width="1546" alt="Screenshot 2020-04-23 at 5 28 02 PM" src="https://user-images.githubusercontent.com/670099/80151807-a0eaa480-8588-11ea-967c-7e744125ba0e.png">
<img width="1546" alt="Screenshot 2020-04-23 at 5 28 18 PM" src="https://user-images.githubusercontent.com/670099/80151814-a2b46800-8588-11ea-9b5d-7d69c3d2f220.png">

Runtime performance seems to be well under 10ms even on the large template (done with 100 consecutive parses, but the time even includes loading the file from disc).
```
bash-3.2$ time ./bin/index.js parse --template ./test/data/templateLarge.json --input ./test/data/large.md 
<plate ./test/data/templateLarge.json --input ./test/data/large.md 
5:19:33 PM - info:
{
  "$class": "org.test.MyContract",
  "clause3": {
    "$class": "org.test.MyClause",
    "seller": "Steve",
    "buyer": "Betty",
    "amount": 3131,
    "currency": "EUR",
    "forceMajeure": false
  },
  "penalty": 10.99
}
Average time:  14.012748956680298
5:19:34 PM - info:
{
  "$class": "org.test.MyContract",
  "clause3": {
    "$class": "org.test.MyClause",
    "seller": "Steve",
    "buyer": "Betty",
    "amount": 3131,
    "currency": "EUR",
    "forceMajeure": false
  },
  "penalty": 10.99
}
Average time:  8.532179951667786
...
5:20:07 PM - info:
{
  "$class": "org.test.MyContract",
  "clause3": {
    "$class": "org.test.MyClause",
    "seller": "Steve",
    "buyer": "Betty",
    "amount": 3131,
    "currency": "EUR",
    "forceMajeure": false
  },
  "penalty": 10.99
}
Average time:  3.1058295691013336
```

**Comment by jeromesimeon** [comment]
> _Warning: this is a test done in isolation for the parser alone outside of Cicero_
> 
> Pretty good news on the memory usage front (and I believe excellent news on the runtime performances).

Instrumented version of `./bin/index.js` used for those tests. 
[index.txt](https://github.com/accordproject/cicero/files/4525771/index.txt)

**Comment by jeromesimeon** [comment]
Some possible additional questions or work item (edited from feedback during review with @dselman @irmerk @DianaLease )

- [ ] Move the text generation closer to the parsing?
- [ ] Parse templates themselves using the same parser technology
- [ ] Can we generate a BNF? Or some lang. independent description of the grammar?
- [ ] Can we make the parsing table extensible by users?
- [ ] Decide whether this should live in markdown transform or not
- [ ] Can we also redo the Concerto parser using the same parser technology?

**Comment by jeromesimeon** [comment]
Another useful benchmarking point, courtesy of @mttrbrts 

https://sap.github.io/chevrotain/performance/

<img width="1445" alt="Screenshot 2020-05-06 at 5 33 03 PM" src="https://user-images.githubusercontent.com/670099/81230564-bc5da280-8fbf-11ea-933d-f1d735294213.png">

**Comment by jeromesimeon** [comment]
Template checklist:
- [ ] acceptance-of-delivery
- [ ] car-rental-tr
- [ ] certificate-of-incorporation
- [ ] copyright-license
- [ ] demandforecast
- [ ] docusign-connect
- [ ] docusign-po-failure
- [ ] eat-apples
- [ ] empty
- [ ] empty-contract
- [ ] fixed-interests
- [ ] fixed-interests-static
- [ ] fragile-goods
- [ ] full-payment-upon-demand
- [ ] full-payment-upon-signature
- [ ] hellomodule
- [ ] helloworld
- [ ] helloworldstate
- [ ] installment-sale
- [ ] interest-rate-swap
- [ ] ip-payment
- [ ] latedeliveryandpenalty
- [ ] latedeliveryandpenalty-else
- [ ] lateinvoicewithpayment
- [ ] log
- [ ] minilatedeliveryandpenalty
- [ ] minilatedeliveryandpenalty-capped
- [ ] minilatedeliveryandpenalty-payment
- [ ] one-time-payment-tr
- [ ] online-payment-contract-tr
- [ ] payment-upon-delivery
- [ ] payment-upon-iot
- [ ] payment-upon-signature
- [ ] perishable-goods
- [ ] promissory-note
- [ ] promissory-note-md
- [ ] rental-deposit
- [ ] rental-deposit-with
- [ ] saft
- [ ] safte
- [ ] sales-contract-ru
- [ ] servicelevelagreement
- [ ] simplelatedeliveryandpenalty
- [ ] supplyagreement
- [ ] supplyagreement-perishable-goods
- [ ] volumediscount
- [ ] volumediscountolist
- [ ] volumediscountulist

**Comment by jeromesimeon** [comment]
### Related issues

- [x] #4: Repeated bindings
- [ ]  #17 #549: Optional parameters
- [x] #386: Remove use of Eval 
- [ ] https://github.com/accordproject/cicero/issues/552: parsing for relationships
- [ ] https://github.com/accordproject/cicero/issues/553: Parser modularity 
- [x] https://github.com/accordproject/cicero/issues/529: Template with no logic

**Comment by jeromesimeon** [comment]
This issue is resolved with the release of Cicero `0.21`. We can open new issues if needed for features that didn't make the cut, or for any bug or new requirements.
